### PR TITLE
os-release: Add VERSION_CODENAME field

### PR DESCRIPTION
**Pull request by bdrung**
Debian and their derivatives (Ubuntu, Trisquel, etc.) use a code name
for their repositories. Thus record the code name in os-release for
processing.

Closes systemd/systemd#3429

### CODE REVIEW

File: man/os-release.xml [hunk 1]
```diff
@@ -309,6 +309,22 @@
...
+      <varlistentry>
+        <term><varname>CODENAME=</varname></term>
+
+        <listitem><para>
+        A lower-case string (no spaces or other characters outside of
+        0–9, a–z, ".", "_" and "-") identifying the operating system
+        release code name, excluding any OS name information or
+        release version, and suitable for processing by scripts or
+        usage in generated filenames. This field is optional and may
+        not be implemented on all systems.
```
Comment (poettering) [inline man/os-release.xml]: Shouldn't this be called VERSION_CODENAME= or so? To match VERSION= and VERSION_ID=?

Comment (martinpitt) [inline man/os-release.xml]: I suppose that's up for bikeshedding; `VERSION_CODENAME=` makes some sense as the other fields like `ID=` or `SUPPORT_URL=` are not specific to the OS version. But then again, `PRETTY_NAME=` _is_ specific to the OS version, so it wouldn't be consistent either way.

Still, I agree that `VERSION_CODENAME=` is a bit more consistent indeed.

**Comment by poettering** [comment]
@bdrung ok, could you please rename the variable, as both @martinpitt and I prefer VERSION_CODENAME?

**Comment by johannbg** [comment]
@poettering since you seem to be so eager to add this entry what will the difference be on version and version codename?

**Comment by poettering** [comment]
@johannbg I am not particularly eager, I am just happy to review the patch. That's all. I'll let @bdrung or @martinpitt answer your question.

**Comment by martinpitt** [comment]
`VERSION=` is a human readable string, see `man os-release`. I. e. it's something like `14.04.4 LTS, Trusty Tahr`. But the corresponding code name that is being used on the archive servers is the machine readable, single-word `trusty`.

Correspondingly on Debian, `VERSION="8 (jessie)`, while `VERSION_ID="8"` and the code name would just be `jessie`.

On F23, `VERSION="23 (Twenty Three)"`. Apparently Fedora stopped doing code names with F20 ("Heisenbug"), so they don't have this field any more; supposedly Fedora also doesn't use the code names on their package servers, Debian-based releases do.

So in summary, `VERSION=` isn't machine parseable as it does not have a strictly defined format, and the code name is nowhere else.

**Comment by johannbg** [comment]
@martinpitt I would not use Fedora as a good sample with regards to the structure of os-release and if this is going to be introduced then this should be introduced as Build="" which is available ( and I have seen in os-release in the wild ) ,generic, clear enough ( unlike the vendor line ) and could include Alpha,Beta,Final,Development,Staging, Master,Production,Official. CODENAME,Branch etc. 
Example
Build="Trusty"
BuildID=12345

Codename has nothing to do with "Version" thus calling this VERSION_CODENAME is confusing at best. 

The reference to codename in the "VERSION=" field in the man page should be removed when this gets committed to avoid any confusion and vendors/distribution directed to use Build="" instead.

**Comment by bdrung** [comment]
The codename is just a name for the version number. It is not connection to a specific build.

For example, after the release of Ubuntu 16.04, the codename yakkety was selected for the next release. It is currently in development and will pass alpha and beta releases. Once it is stable, it will be released as Ubuntu 16.10, but it will keep its codename yakkety.

On Debian, the codename sid will always stay the development version. testing is currently called stretch and it will be released as Debian 9 in the future.

The codename is important for Debian/Ubuntu, because it is used in the package repository.

**Comment by bdrung** [comment]
Here you go. The field is called VERSION_CODENAME now.

**Comment by johannbg** [comment]
@poettering  I'll file a separate patch for the Build= which can include what I mentioned above ( Including Codenames ) and can be relied upon unlike VERSION field where it's clearly stated in the man page for rolling releases applications should not rely on these fields to be set hence codename like Rawhide and Tumbleweed etc can be used there.

**Comment by poettering** [comment]
@johannbg there's already BUILD_ID=. Please understand that I am much more inclined to add new fields here, if it's established distros who ask for them and can make a good case. Adding random fields that no distribution actually benefits from is not a priority.

**Comment by johannbg** [comment]
@poettering for the first I was proposing "BUILD" not "BUILD_ID" and who asks for it should be irrelevant let alone an established distro that actually does not make a good usecase for it ( at least from my pov ) and I thought that was clear BUILD is already being used by other distribution ( with presiding <distribution> tag ) and it suits better instead of using some out of context name like "VERSION_CODENAME" which directly conflicts with what VERSION is already being used for and and is either unusable or causes confusion for rolling releases due to the notes in the man page I refer to in that comment.  And FYI you need start extending your distribution scope further than just "Fedora/Redhat" and "Debian/Ubuntu." and view things more as an neutralized upstream when making decisions like this ( let alone implementing things based on the needs of one distribution or it's derivatives Debian/Ubuntu or Fedora/Redhat ). If you take your time and do some research outside that limited scope of yours, you will see that "BUILD" is more generic enough and would cover quite more usecases than just "CODENAME". Both ofcourse can coexist but then people will just use VERSION ( Like they do now ), VERSION_CODENAME ( which is what being proposed ) and BUILD ( which is already being used to implement codename along with release stages as in alpha,beta,final,  deviation between community releases and enterprise release, various stages of development like staging,production etc and other tags ) to implement CODENAME

**Comment by johannbg** [comment]
@poettering  Here's an example of an ( embedded ) distribution [1] that uses BUILD ( along with ARCH so that also might be common on at least embedded platforms to deviate which build belongs to which architecture  ) 

So you can see BUILD fits embedded/server/containers/image development/deployment of anykind just fine as I mentioned thus should be more widely adopted rather than VERSION_CODENAME.  
1. https://github.com/OpenELEC/OpenELEC.tv/blob/master/scripts/image

**Comment by poettering** [comment]
what openelec is doing there with OPENELEC_BUILD might be something for VARIANT= I think.

**Comment by johannbg** [comment]
No Variant is more like VARIANT="LTS"  and in the case of embedded ( and theirs ) variant would be VARIANT="Rasperry PI", VARIANT="Freescale IMX6" since their variant is device spesific.  Deviation between an "community" release and "enterprise" release ( which they seem to be calling "official" there ) belongs in "BUILD"

What OpenElec is doing there ( and probably most embedded distributions ) is using on one hand ARCH= as in ARCH=ARMv7 or ARCH=RPI2.ARMv7 ( the former part here should belong in variant if that's the case ) and BUILD= for the codename or the buildname of the image then build_id is used for either for the release ID or checksum of that build image. In anycase BUILD is useful and used in the wild for wide variety of purpose which includes codename. 

Just for "image" handling ( distribution releases/container images ) this probably would makes sense to be implemented in os-release
ARCH and or BUILD_ARCH, BUILD, BUILD_ID and probably BUILDSUM ( to make the distinction from build_id ).
